ml/utils.py

```diff
@@ -0,0 +1,400 @@
+    tdf = get_nearest(search_term, src, dst, edf)
+    # now find all their nearest connections.
+    if both:
+        gcols = pd.concat([tdf[dst], tdf[src]], axis=0)
+        logger.info(
+            f"Then finding all edges from {search_term} in {src} and {dst} columns of full edgeDataFrame"
+        )
+        df = edf[(edf[src].isin(gcols) | edf[dst].isin(gcols))]
+    else:
+        # logger.info(f'Finding {search_term} in {src} columns')
+        # tdf = search_to_df(search_term, src, edf)
+        logger.info(
+            f"Then finding {src} columns with edges from {search_term} in {dst} column of full edgeDataFrame"
+        )
+        df = edf[edf[src].isin(list(tdf[dst]) + [search_term])]
+    return df
+
+
+def get_graphistry_from_milieu_search(
+    search_term, src, dst, node_col, edf, ndf, both=False
+):
+    """
+        Produces large graphs of neighbors from a given search term
+    :param search_term:
+    :param src:
+    :param dst:
+    :param node_col:
+    :param edf:
+    :param ndf:
+    :param both:
+    :return:
+    """
+    tdf = get_milieu_graph_from_search(search_term, src, dst, edf, both=both)
+    gcols = pd.concat(
+        [tdf[dst], tdf[src]], axis=0
+    )  # get all node entries that show up in edge graph
+    ntdf = ndf[ndf[node_col].isin(gcols)]
+    g = graphistry.edges(tdf, src, dst).nodes(ntdf, node_col)
+    return g
+
+
+# #################################################################################################
+#
+#   DGL helpers
+#
+# #################################################################################################
+
+
+def get_available_devices():
+    """Get IDs of all available GPUs.
+
+    Returns:
+        device (torch.device): Main device (GPU 0 or CPU).
+        gpu_ids (list): List of IDs of all GPUs that are available.
+    """
+    gpu_ids = []
+    if torch.cuda.is_available():
+        gpu_ids += [gpu_id for gpu_id in range(torch.cuda.device_count())]
+        device = torch.device(f"cuda:{gpu_ids[0]}")
+        torch.cuda.set_device(device)
+    else:
+        device = torch.device("cpu")
+
+    return device, gpu_ids
+
+
+def reindex_edgelist(df, src, dst):
+    """Since DGL needs integer contiguous node labels, this relabels as preprocessing step
+    :eg
+        df, ordered_nodes_dict = reindex_edgelist(df, 'to_node', 'from_node')
+    :param df: edge dataFrame
+    :param src: source column of dataframe
+    :param dst: destination column of dataframe
+
+    :returns
+        df, pandas DataFrame with new edges.
+        ordered_nodes_dict, dict ordered from most common src and dst nodes.
+    """
+    srclist = df[src]
+    dstlist = df[dst]
+    cnt = Counter(pd.concat([srclist, dstlist], axis=0))
+    ordered_nodes_dict = {k: i for i, (k, c) in enumerate(cnt.most_common())}
+    df[config.SRC] = df[src].apply(lambda x: ordered_nodes_dict[x])
+    df[config.DST] = df[dst].apply(lambda x: ordered_nodes_dict[x])
```

Comment:
vectorize via join?

Reply:
cuGraph does this internally and exposes some methods, but that part of the API is unstable, so probably stay away from